# Concatenate tables


The main porpose of this script is to concatenate several tables with the same format into one. 

Fisrt, let's check what we have:

In [1]:
import os
import pandas as pd


In [2]:
files = os.listdir('proteins')
print('The folder "proteins" has %d files' % len(files))
extensions = set()
for file in files:
    ext = file.split('.')[-1]
    extensions.add(ext)
    
print("The files have the follwing extensions:", ','.join(extensions))

The folder "proteins" has 632 files
The files have the follwing extensions: xls


So we have 632 files in .xls format. 

Despite of the .xls format, the files are text files separated by tab.

In [3]:
pd.read_table('proteins/gi_110758919.xls', sep='\t').head()

Experiment name Biological sample category Biological sample name  \
0  TUB Arachinidar e pep go         Shotgun Proteomics                    TUB   
1  TUB Arachinidar e pep go         Shotgun Proteomics                    TUB   
2  TUB Arachinidar e pep go         Shotgun Proteomics                    TUB   

                      MS/MS sample name  \
0  Mudpit_DATA.TXT (F003448 TUB NCBInr)   
1  Mudpit_DATA.TXT (F003448 TUB NCBInr)   
2  Mudpit_DATA.TXT (F003448 TUB NCBInr)   

                                       Protein group Protein accession number  \
0  PREDICTED: constitutive coactivator of PPAR-ga...             gi|110758919   
1  PREDICTED: constitutive coactivator of PPAR-ga...             gi|110758919   
2  PREDICTED: constitutive coactivator of PPAR-ga...             gi|110758919   

                                        Protein name  \
0  PREDICTED: constitutive coactivator of PPAR-ga...   
1  PREDICTED: constitutive coactivator of PPAR-ga...   
2  PREDICTED: constitutive coactivator of PPAR-ga...   

  Protein identification probability Protein percentage of total spectra  \
0                             100.0%                            0.00131%   
1                             100.0%                            0.00131%   
2                             100.0%                            0.00131%   

  Protein molecular weight (AMU)         ...          Mascot Ion score  \
0                      109,122.2         ...                      12.5   
1                      109,122.2         ...                      13.0   
2                      109,122.2         ...                      12.5   

   Mascot Identity score  Mascot Delta Ion score  \
0                   36.7                    0.00   
1                   36.8                   -3.02   
2                   36.1                   -0.71   

  Modifications identified by spectrum Actual peptide mass (AMU)  \
0                Carbamidomethyl (+57)                  2,035.73   
1                      Oxidation (+16)                    996.53   
2                      Oxidation (+16)                  3,382.33   

  Spectrum charge Actual minus calculated peptide mass (AMU)  \
0               3                                      -0.13   
1               2                                   -1000.00   
2               4                                     850.00   

  Actual minus calculated peptide mass (PPM)  Peptide start index  \
0                                        -64                   63   
1                                   -1000000                  458   
2                                     250000                  458   

   Peptide stop index  
0                  79  
1                 474  
2                 479  

[3 rows x 28 columns]

Time to concatenate all files into one to make future analysis easier. However, a new column (file_name) will be defined to know the origin of the row.

In [4]:
def parse_protein_table(fname):
    df = pd.read_table('proteins/{}'.format(fname))
    df['file_name'] = fname
    return df
    

In [5]:
%%time
df = parse_protein_table(files[0])

for f in files[1:]:
    df = pd.concat([df, parse_protein_table(f)])

CPU times: user 5.64 s, sys: 28 ms, total: 5.66 s
Wall time: 5.66 s


## Thanks, Python! <3

Wow! Within ten seconds we solve a problem that would be a very boring task.

In [6]:
df.shape

(2214, 29)

In [7]:
df.to_csv('all_proteins.tsv', sep='\t', index=None)